### PR TITLE
[Calyx] Add CalyxToHW scaffolding.

### PR DESCRIPTION
**Pull request by cgyurgyik**

### CODE REVIEW

File: lib/Conversion/CalyxToSV/CalyxToSV.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//
+// This is the main Calyx to SV Conversion Pass Implementation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Conversion/CalyxToSV/CalyxToSV.h"
+#include "../PassDetail.h"
+#include "circt/Dialect/Calyx/CalyxOps.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "mlir/Transforms/DialectConversion.h"
```
Comment (mikeurbach) [inline lib/Conversion/CalyxToSV/CalyxToSV.cpp:17]: Have you thought about how you'd use the dialect conversion framework here? I'm a fan of this, but it does require a different style of pass compared to how the other Calyx passes work. It may be a good tool for this conversion, so I'm not trying to dissuade you. But you don't have to use it just because you're doing a "conversion". If you already have an algorithm in mind, it is sometimes easier to implement it using the APIs you've already worked with rather than the dialect conversion framework.

File: lib/Conversion/CalyxToSV/CalyxToSV.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+public:
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+void CalyxToSVPass::runOnOperation() {
+  // auto op = getOperation();
+  ConversionTarget target(getContext());
+  target.addIllegalDialect<CalyxDialect>();
+  target.addLegalDialect<SVDialect>();
```
Comment (mikeurbach) [inline lib/Conversion/CalyxToSV/CalyxToSV.cpp:36]: Re-iterating previous comments, I'd expect you'd want the HW and Comb dialects here, if using the dialect conversion framework.

**Review by mikeurbach — COMMENTED** [review]
The scaffolding looks good in general, modulo some implementation details I'm curious about.

File: lib/Conversion/CalyxToSV/CalyxToSV.cpp (hunk 1 above)
Comment (cgyurgyik) [reply lib/Conversion/CalyxToSV/CalyxToSV.cpp:17]: To be frank, this was the result of a copy-and-paste :-). I'm not too familiar with the dialect conversion framework, so I'll have to do some reading. Do you have any concrete examples you'd suggest looking at in CIRCT, for both using / not using it?

File: include/circt/Conversion/Passes.td [hunk 3]
```diff
@@ -15,6 +15,19 @@
...
+// CalyxToSV
+//===----------------------------------------------------------------------===//
+
+def CalyxToSV : Pass<"lower-calyx-to-sv", "mlir::ModuleOp"> {
+  let summary = "Lower Calyx to SV";
+  let description = [{
+    This pass lowers Calyx to SV.
+  }];
+  let constructor = "circt::createCalyxToSVPass()";
+  let dependentDialects = ["sv::SVDialect"];
```
[the remark this answers is not in the document]
Comment (cgyurgyik) [reply include/circt/Conversion/Passes.td:28]: Ok thanks. This makes sense.

As far as generating Comb ops, I don't think so. When Calyx is fully lowered (i.e. control is compiled), all wires should be appropriately guarded.

File: lib/Conversion/CalyxToSV/CalyxToSV.cpp (hunk 1 above)
Comment (cgyurgyik) [reply lib/Conversion/CalyxToSV/CalyxToSV.cpp:17]: Looks like FIRRTL's [LowerToHW](https://github.com/llvm/circt/blob/main/lib/Conversion/FIRRTLToHW/LowerToHW.cpp) doesn't use it, while [HandshakeToFIRRTL](https://github.com/llvm/circt/blob/main/lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp) does. It seems OK to follow FIRRTL's approach. It isn't immediately clear to me what the benefits of using a [conversion framework](https://mlir.llvm.org/docs/DialectConversion/) are.

Comment (mikeurbach) [reply lib/Conversion/CalyxToSV/CalyxToSV.cpp]: I think the HandshakeToFIRRTL isn't the greatest example, since it doesn't actually use the framework in an interesting way, it fully converts handshake.func and its body into a firrtl.module, and has a giant monolithic pattern with lots of logic. The idea of the framework is to have much smaller patterns that locally convert each op you care about. You also have the TypeConverter functionality, to bridge between different type systems (if the IR you're converting has different type systems). An example in CIRCT that is trying to follow the best practices (see [this talk](https://drive.google.com/file/d/1VfVajitgf8ZPnd-HRkJvaJiFLhBsluXN/view?usp=sharing) for more details) is here: https://github.com/llvm/circt/blob/main/lib/Conversion/HWToLLHD/HWToLLHD.cpp.

In short, I think if you just want to follow what the existing Calyx compiler does, you will be better off writing your own pass without the dialect conversion framework, like FIRRTL does. If you want to take this opportunity to do something that's new and MLIR-specific, the dialect conversion framework can be a powerful tool, but there's a bit of a learning curve. I'd start with the above talk and look at the HWToLLHD conversion first.

**Review by mikeurbach — APPROVED** [review]
